firestore/lib/src/data/restaurant_provider.dart

```diff
@@ -0,0 +1,168 @@
+/*
+ *  Copyright 2022 Google LLC
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+import 'dart:async';
+
+import 'package:cloud_firestore/cloud_firestore.dart';
+
+import '../model/filter.dart';
+import '../model/restaurant.dart';
+import '../model/review.dart';
+
+abstract class RestaurantProvider {
+  Stream<List<Restaurant>> get allRestaurants;
+  Future<void> addRestaurant(Restaurant restaurant);
+  Future<void> addReview(
+      {required String restaurantId, required Review review});
+  void addRestaurantsBatch(List<Restaurant> restaurants);
+  void loadAllRestaurants();
+  void loadFilteredRestaurants(Filter filter);
+  Future<Restaurant> getRestaurantById(String restaurantId);
+  void dispose();
+}
+
+class FirestoreRestaurantProvider implements RestaurantProvider {
+  FirestoreRestaurantProvider() {
+    allRestaurants = _allRestaurantsController.stream;
+  }
+
+  final StreamController<List<Restaurant>> _allRestaurantsController =
+      StreamController();
+
+  @override
+  late final Stream<List<Restaurant>> allRestaurants;
+
+  @override
+  Future<void> addRestaurant(Restaurant restaurant) {
+    final restaurants = FirebaseFirestore.instance.collection('restaurants');
+    return restaurants.add({
+      'avgRating': restaurant.avgRating,
+      'category': restaurant.category,
+      'city': restaurant.city,
+      'name': restaurant.name,
+      'numRatings': restaurant.numRatings,
+      'photo': restaurant.photo,
+      'price': restaurant.price,
+    });
+  }
+
+  @override
+  void addRestaurantsBatch(List<Restaurant> restaurants) {
+    for (var restaurant in restaurants) addRestaurant(restaurant);
+  }
+
+  @override
+  Future<void> addReview(
+      {required String restaurantId, required Review review}) async {
+    final restaurant = await FirebaseFirestore.instance
```

Comment:
Writing testable Firebase code is so tricky. This pattern is a good start, because at least the rest of this app will be able to easily mock Firebase, _but_, this function itself remains nearly impossible to test.

The folks at Very Good Ventures have done a lot of thinking about this, and we should chat in the future. A huge value add we could forward to the community is spreading their patterns for getting up to 100% test coverage on Firestore code.

(Whether or not one actually wants strict 100% coverage is another topic 😆 )

Reply:
This is a great statement for me to read,  as I thought about how to test for a long time and then eventually just forgot to follow through. I'll talk to VG,V Puf, etc for insights